components/TextFields/src/ContainedInputView/MDCBaseTextFieldLayout.m

```diff
@@ -0,0 +1,120 @@
+
+static const CGFloat kHorizontalPadding = (CGFloat)12.0;
+
+/**
+ This constant can go away once the code that determines container height is brought over.
+ */
+static const CGFloat kTemporaryVerticalPadding = (CGFloat)12.0;
+
+@interface MDCBaseTextFieldLayout ()
+@end
+
+@implementation MDCBaseTextFieldLayout
+
+#pragma mark Object Lifecycle
+
+- (instancetype)initWithTextFieldSize:(CGSize)textFieldSize
+                             leftView:(UIView *)leftView
+                         leftViewMode:(UITextFieldViewMode)leftViewMode
+                            rightView:(UIView *)rightView
+                        rightViewMode:(UITextFieldViewMode)rightViewMode
+                                isRTL:(BOOL)isRTL
+                            isEditing:(BOOL)isEditing {
+  self = [super init];
+  if (self) {
+    [self calculateLayoutWithTextFieldSize:textFieldSize
+                                  leftView:leftView
+                              leftViewMode:leftViewMode
+                                 rightView:rightView
+                             rightViewMode:rightViewMode
+                                     isRTL:isRTL
+                                 isEditing:isEditing];
+    return self;
+  }
+  return nil;
+}
+
+#pragma mark Layout Calculation
+
+- (void)calculateLayoutWithTextFieldSize:(CGSize)textFieldSize
+                                leftView:(UIView *)leftView
+                            leftViewMode:(UITextFieldViewMode)leftViewMode
+                               rightView:(UIView *)rightView
+                           rightViewMode:(UITextFieldViewMode)rightViewMode
+                                   isRTL:(BOOL)isRTL
+                               isEditing:(BOOL)isEditing {
+  BOOL shouldAttemptToDisplayLeftView = [self shouldAttemptToDisplaySideView:leftView
+                                                                    viewMode:leftViewMode
+                                                                   isEditing:isEditing];
+  BOOL shouldAttemptToDisplayRightView = [self shouldAttemptToDisplaySideView:rightView
+                                                                     viewMode:rightViewMode
+                                                                    isEditing:isEditing];
+  CGFloat leftViewWidth = CGRectGetWidth(leftView.frame);
+  CGFloat leftViewHeight = CGRectGetHeight(leftView.frame);
+  CGFloat leftViewMinX = 0;
+  CGFloat leftViewMinY = 0;
+  if (shouldAttemptToDisplayLeftView) {
+    leftViewMinX = kHorizontalPadding;
+    leftViewMinY = kTemporaryVerticalPadding;
+  }
+
+  CGFloat rightViewWidth = CGRectGetWidth(rightView.frame);
+  CGFloat rightViewHeight = CGRectGetHeight(rightView.frame);
+  CGFloat rightViewMinX = 0;
+  CGFloat rightViewMinY = 0;
+  if (shouldAttemptToDisplayRightView) {
+    CGFloat rightViewMaxX = textFieldSize.width - kHorizontalPadding;
+    rightViewMinX = rightViewMaxX - rightViewWidth;
+    rightViewMinY = kTemporaryVerticalPadding;
+  }
```

Comment:
Is it possible to remove the conditional for the origin and rely only on the hidden property?

Suggestion: if so, you might be able to simplify this code to something like so:

```objc
static const UIEdgeInsets kContentPadding = {12, 12, 12, 12};

CGRect bounds = CGRectStandardize(CGRectMake(0, 0, textFieldSize.width, textFieldSize.height));
CGRect contentBounds = UIEdgeInsetsInsetRect(contentFrame, kContentPadding);

CGRect leftViewFrame = CGRectStandardize(leftView.frame);
leftViewFrame.origin.x = paddedContentFrame.origin.x;
leftViewFrame.origin.y = paddedContentFrame.origin.y;

CGRect rightViewFrame = CGRectStandardize(rightView.frame);
rightViewFrame.origin.x = CGRectGetMaxX(paddedContentFrame) - rightViewFrame.size.width;
rightViewFrame.origin.y = paddedContentFrame.origin.y;
```

Reply:
I'm hesitant to use a CGRect inset from the text field bounds to determine the min/max Xs and Ys of the subviews because these values depend on so many things, and are treated differently depending on whether we're talking about vertical or horizontal positioning. The logic behind calculating horizontal positions of views is pretty straightforward, and can be done making use of CGFloat constants for padding. The vertical positions of views depend on things like state, labelBehavior, and density.

Reply:
*nods* we don't have an established pattern in our code, but making use of `UIEdgeInsetsInsetRect` can be a helpful way to semantically model the relative layout considerations of manual layout code using a padding/margins-like approach. Changing to these APIs changes your code from roughly being "working with a bunch of independent CGFloat values" to "working with frames and their relative positioning to one another". CGRectDivide is another very helpful API when it comes to dividing up an interface into relevant semantic portions.

This isn't blocking feedback, but I would appreciate if you took it into consideration for future code.